### DISCUSSION TITLE
Can I set custom cookies when successfully logged in?

### DISCUSSION THREAD

**Discussion by dipak-webo** [nextauthjs/next-auth#4428] [Help] [answered]
### Question 💬

Hi,

I've been using Next-Auth CredentialsProvider, I want to set some custom cookies when logged in, but it didn't work.

Login works perfectly but it doesn't set custom cookies. My Next-Auth version is `4.2.1`. Is there any way to set custom cookies?

### How to reproduce ☕️

```js
export default (req, res) => {
  return NextAuth(req, res, {
    providers: [
      CredentialsProvider({
        name: "creds",

        credentials: {
          username: { label: "Username", type: "text", autoComlete: 'off' },
          password: {  label: "Password", type: "password" }
        },

        async authorize(credentials) {
          const user = await login(credentials);

          if ( user ) {
             res.setHeader('Set-Cookie', 'my_cookie=value');
             return user;
          }

          throw new Error("Invalid creds");
        },
      }),
    ],
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ThangHuuVu** [comment]
I don't think there is an API for setting custom cookies in NextAuth.js atm 🤔 cc @balazsorban44. Setting cookies in the `authorize` callback here won't work. Instead, could you try checking the session on `getServerSideProps` and set cookies there?

**Reply by OxiBo** [reply]
Why won't setting cookie work in authorize? According to one of the answers here it should be possible - https://stackoverflow.com/questions/68235182/nextjs-with-next-auth-setting-cookie-received-from-node-js

**Comment by balazsorban44** [comment]
I would like to hear the use case for this.

**Reply by dipak-webo** [reply]
Hi @balazsorban44 

I am using Next-Auth to authenticate users with WordPress on the backend, so when successfully logged in, need to set WordPress cookies as well.

**Comment by dipak-webo** [comment]
Hi Guys,

The package `nookies` allowed to set the custom cookies. Thanks.

```js
import { setCookie } from 'nookies';

 async authorize(credentials){
  //...
  setCookie({ res }, 'name', 'value',{
     maxAge: 2 * 24 * 60 * 60,
     path: '/',
     httpOnly: true
  })
}
```

**Reply by rnnyrk** [reply]
Where do you get `res` from? Besides `credentials` the `authorize` function seems only to return `req` ?

<img width="1443" alt="Screenshot 2022-09-12 at 18 11 52" src="https://user-images.githubusercontent.com/4519142/189703771-04e9fa90-efee-4943-af44-68f423a322c5.png">

**Reply by tiriana** [reply]
I think you need to export a function returning NextAuth, not NextAuth itself. 

Take a look at this: https://stackoverflow.com/questions/67594977/how-to-send-httponly-cookies-client-side-when-using-next-auth-credentials-provid/69418553#69418553

**Reply by gride29** [reply, marked as answer]
That works, if you are having problems with TypeScript types, simply change **[...nextauth].ts** file to **[...nextauth].js**.

Example working code that saves cookies:

```
import axios from "axios";
import NextAuth from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import GithubProvider from "next-auth/providers/github";
import GoogleProvider from "next-auth/providers/google";
import { setCookie } from "nookies";

const nextAuthOptions = (req, res) => {
	return {
		providers: [
			GithubProvider({
				clientId: process.env.GITHUB_ID,
				clientSecret: process.env.GITHUB_SECRET,
			}),
			GoogleProvider({
				clientId: process.env.GOOGLE_CLIENT_ID,
				clientSecret: process.env.GOOGLE_CLIENT_SECRET,
			}),
			CredentialsProvider({
				name: "credentials",
				credentials: {
					username: { label: "username", type: "text" },
					password: { label: "password", type: "password" },
				},
				async authorize(credentials) {
					if (!credentials?.username || !credentials?.password) {
						throw new Error("Invalid credentials");
					}

					return axios
						.post("http://localhost:8080/api/auth/signin", {
							username: credentials.username,
							password: credentials.password,
						})
						.then((response) => {
							const user = response.data;
							setCookie({ res }, "user", JSON.stringify(user), {
								maxAge: 2 * 24 * 60 * 60,
								path: "/",
								httpOnly: true,
							});
							return user;
						})
						.catch((error) => {
							console.log(error);
						});
				},
			}),
		],
		pages: {
			signIn: "/",
		},
		debug: process.env.NODE_ENV === "development",
		session: {
			strategy: "jwt",
		},
		secret: process.env.NEXTAUTH_SECRET,
	};
};

export default (req, res) => {
	return NextAuth(req, res, nextAuthOptions(req, res));
};
```

**Reply by DoronTorangy** [reply]
Is there is a similar solution for next 13 with app directory for [Route handlers](https://nextjs.org/docs/app/building-your-application/routing/router-handlers)?

currently I'm defining next auth like so:

```ts
// app/api/auth/[...nextauth]/route.ts

export const authOptions: AuthOptions = {
    ...
}

const handler = NextAuth(authOptions)
export { handler as GET, handler as POST }
```

I do have access to the request object, but not the response object, since its not passed (need to be manually constructed with `NextResponse.json(...)`

**Reply by gride29** [reply]
@DoronTorangy I believe that you need to extract the **response object** in some way to be able to set cookies.  That was the case at least in my project and I am also using Next 13.

**Reply by DoronTorangy** [reply]
How did you do that? Because your solution uses the pages API

**Reply by gride29** [reply]
Yes indeed it uses pages API. I've created [...nextauth].js file in /pages/api/auth and the content of the file is basically the code I've posted above. 

Below is the code that calls next-auth signIn function. 

```
const onSubmit: SubmitHandler<FieldValues> = (data) => {
    setIsLoading(true);
    
    signIn("credentials", {
	    ...data,
	    redirect: false,
    }).then((callback) => {
	    setIsLoading(false);
    
	    if (callback?.ok) {
		    toast.success("Logged in");
		    router.refresh();
		    loginModal.onClose();
	    }
    
	    if (callback?.error) {
		    toast.error(callback.error);
	    }
    });
};
```

This is how I get the currently logged user: 
```
import { cookies } from "next/headers";
import axios from "axios";
import { signOut } from "next-auth/react";
import Router from "next/navigation";

const parseJwt = (token: string) => {
	try {
		return JSON.parse(atob(token.split(".")[1]));
	} catch (e) {
		return null;
	}
};

export default async function getCurrentUser() {
	// Retrieve user from cookies
	const cookieStore = cookies();
	let user = JSON.parse(cookieStore.get("user")?.value || "null");

	if (user) {
		const decodedJwt = parseJwt(user.accessToken);

		// If the token is expired, we sign out the user
		if (decodedJwt.exp * 1000 < Date.now()) {
			return undefined;
		}
	}

	if (user == null) {
		// If user is null, we check if the user is logged in with OAuth
		const cookieName = "JSESSIONID";
		const cookieValue = cookies().get(cookieName)?.value || null;

		if (cookieValue == null) {
			return undefined;
		}

		return axios
			.get("http://localhost:8080/api/auth/oauthUser", {
				headers: {
					Cookie: `${cookieName}=${cookieValue}`,
				},
			})
			.then((response) => {
				user = response.data;
				user.JSESSIONID = cookieValue;
				return user;
			})
			.catch((error) => {
				console.log(error);
			});
	}

	return user;
}
```

And then I just pass the user object to any component I want:
```
export default async function RootLayout({
	children,
}: {
	children: React.ReactNode;
}) {
	const user = await getCurrentUser();

	return (
		<html lang="en">
			<body className={font.className}>
				<ClientOnly>
					<ToasterProvider />
					<RentModal currentUser={user} />
					<LoginModal />
					<RegisterModal />
					<Navbar currentUser={user} />
				</ClientOnly>
				<div className="pb-20 pt-28">{children}</div>
			</body>
		</html>
	);
}
```

One thing to keep in mind, I am not using **only** Next-Auth as my auth service. Next-Auth in my project acts as some kind of middleware that calls Spring Boot backend, which handles login logic and returns JWT token if credentials are correct, which then I store in the **cookie**. Hope this helps!

**Reply by Yetunde495** [reply]
HI, were you able to find a solution? I'm currently facing the same issue

**Reply by Afaqak** [reply]
any luck?

**Reply by jismonthomas** [reply]
If anyone using route handler you can set cookies using 'coookies' from next/headers as below:

`import { cookies } from 'next/headers'; `

```
//use the following wherever you wanna set youw own cookies

cookies().set({
      name: 'nameOfCookie',
      value: 'valueOfCookie,
      httpOnly: true, //optional
      maxAge: timeOut, //optional
  });
```

you can read more about it here: [app router cookies]( https://nextjs.org/docs/app/api-reference/functions/cookies)

**Reply by alexandre-tobia** [reply]
@Afaqak @DoronTorangy You can do something like that:

```js

export async function GET(req, res) {
  return await NextAuth(req, res, authOptions(req, res));
}

export async function POST(req, res) {
  return await NextAuth(req, res, authOptions(req, res));
}

```

**Comment by ezechuka** [comment]
@DoronTorangy were you or anyone else able to find a solution for nextjs 13?

**Reply by gregoire-jianda** [reply]
This works for me with AppRouter :
 
```
// app/api/auth/[...nextauth]/route.ts
import NextAuth from "next-auth";
import { authOptions } from "@/lib/authOptions";
import { NextRequest } from "next/server";

interface RouteHandlerContext {
    params: { nextauth: string[] };
}

const handler = async (req: NextRequest, context: RouteHandlerContext) => {
    return NextAuth(req, context, authOptions);
};

export { handler as GET, handler as POST };
```

```
// @/lib/authOptions.ts
import { NextAuthOptions } from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import axios from "axios";
import { cookies } from "next/headers";
import { parse } from "cookie";

export const authOptions: NextAuthOptions = {
    // Secret for Next-auth, without this JWT encryption/decryption won't work
    secret: process.env.NEXTAUTH_SECRET,

    // Configure one or more authentication providers
    providers: [
        CredentialsProvider({
            name: "Credentials",
            credentials: {
                email: { label: "Email", type: "text" },
                password: { label: "Password", type: "password" },
            },
            async authorize(credentials, req) {
                try {
                    const response = await axios.post(
                        YOUR_AUTH_ENDPOINT,
                        {
                            email: credentials?.email,
                            password: credentials?.password,
                        },
                    );

                    const apiCookies = response.headers["set-cookie"];
                    if (apiCookies && apiCookies.length > 0) {
                        apiCookies.forEach((cookie) => {
                            const parsedCookie = parse(cookie);
                            const [cookieName, cookieValue] =
                                Object.entries(parsedCookie)[0];
                            const httpOnly = cookie.includes("httponly;");

                            cookies().set({
                                name: cookieName,
                                value: cookieValue,
                                httpOnly: httpOnly,
                                maxAge: parseInt(parsedCookie["Max-Age"]),
                                path: parsedCookie.path,
                                sameSite: parsedCookie.samesite,
                                expires: new Date(parsedCookie.expires),
                                secure: true,
                            });
                        });
                    }

                    return response.data;
                } catch (error) {
                    console.log(error);
                    // @ts-expect-error
                    throw Error(error.response);
                }
            },
        }),
    ],
};
```

Note that the `parse` method of the cookie module (https://www.npmjs.com/package/cookie) didn't seem to pick up the httpOnly option, so I checked it separately.

**Reply by ezechuka** [reply]
@gbelorgey, I just tried it now and it worked. Thanks! 
May I ask what's the purpose of RouteHandlerContext? Also can I access the cookie in the api folder

**Reply by gregoire-jianda** [reply]
@ezechuka  Regarding RouteHandlerContext, it's the second argument passed to a dynamic route handler in Next.js : https://nextjs.org/docs/app/building-your-application/routing/route-handlers#dynamic-route-segments.

You can call your API from the server side without even using `cookies().set`, the cookie trick only helps if you want to call it from the client side.

**Reply by microooji** [reply]
@gbelorgey Thank you for code! When I use the cookies().set function as in your example, a session is not created and I do not receive a token. I end up passing the session to layout.ts to `<AuthProvider>`  (`const session = await getServerSession(authOptions);`) but its `null`. Have you encountered this?

**Reply by microooji** [reply]
**SOLVED:** 

I upgraded node v18.12 to node v18.18 and am using these package versions:

`{
    "next": "13.5.3",
    "next-auth": "^4.23.2"
}`

**Comment by jismonthomas** [comment]
> Is there is a similar solution for next 13 with app directory for [Route handlers](https://nextjs.org/docs/app/building-your-application/routing/router-handlers)?
> 
> currently I'm defining next auth like so:
> 
> ```ts
> // app/api/auth/[...nextauth]/route.ts
> 
> export const authOptions: AuthOptions = {
>     ...
> }
> 
> const handler = NextAuth(authOptions)
> export { handler as GET, handler as POST }
> ```
> 
> I do have access to the request object, but not the response object, since its not passed (need to be manually constructed with `NextResponse.json(...)`

If anyone using route handler you can set cookies using 'coookies' from next/headers as below:

`import { cookies } from 'next/headers'; `

```
//use the following wherever you wanna set youw own cookies

cookies().set({
      name: 'nameOfCookie',
      value: 'valueOfCookie,
      httpOnly: true, //optional
      maxAge: timeOut, //optional
  });
```

you can read more about it here: [app router cookies]( https://nextjs.org/docs/app/api-reference/functions/cookies)

**Comment by trangchongcheng** [comment]
app/api/auth/[...nextauth]/route.ts

import { cookies } from 'next/headers'; 

cookies().set({
              name: 'auth',
              value: 'valueOfCookie',
              httpOnly: true,
            })
I working to me, perfect.... 
i using nextjs 13.5

**Comment by AndonMitev** [comment]
guys have you manage to create interceptor and set this cookie for header, like i have some progress but then problem comes that some requests are client other are server and cnat use cookies :@

**Comment by parthtiwar-i** [comment]
I think it worked with credentialProvider but how can we set cookie with an oauth like google
So what I want is i ll use the google signup login of next-auth and  after google signin ill send the provider, providerId and profile to my BE where ill store data and create a cookie and set the cookie to header and sent response.
so that MY BE cookie should be set at the header.
How can i do this anyone ?

**Reply by georgezouq** [reply]
same problem, any solution?
